src/infra/task-system-inmem/src/task_executor_inmem.rs

```diff
@@ -64,6 +76,131 @@ impl TaskExecutorInMemory {
             })
             .await
     }
+
+    async fn make_dataset_update_result(
+        &self,
+        dataset_id: &DatasetID,
+        pull_result: PullResult,
+    ) -> Result<TaskUpdateDatasetResult, InternalError> {
+        Ok(match pull_result {
```

Comment:
Too much nesting here is making code hard to read. I'd suggest an early return on `PullResult::UpToDate`.

Reply:
Split on 2 methods. I've also reduce 1 level by utilizing `std::mem::drop` for the stream to unborrow refs.